File: python/station-distance-py/stations.py
import csv


class Stations:
    """
    The Station class manages the collection of stations and provides some methods for it.
    """
    # List of used keys/columns of the data
    _USED_KEYS: list[str] = ['DS100', 'NAME', 'Laenge', 'Breite']

    # Path to input csv file
    _CSV_PATH: str = 'data/D_Bahnhof_2020_alle.CSV'

    _station_collection: dict = {}

    def __init__(self):
        station_list: list = []
        # Read input csv file
        with open(self._CSV_PATH, 'r') as infile:
            for line in csv.DictReader(infile, delimiter=';'):
                # Filter on only Fernverkehrshalte (FV)
                if line['Verkehr'] == 'FV':
                    # Convert fields Laenge and Breite to floats
                    line['Laenge'] = float(line['Laenge'].replace(',', '.'))
                    line['Breite'] = float(line['Breite'].replace(',', '.'))
                    # Some stations have multiple DS100 codes
                    # Therefore they need to be added individually.
                    ds100codes = line['DS100'].split(',')
                    for code in ds100codes:
                        x: dict = line.copy()
                        x['DS100'] = code
                        # Filter dict to contain only used key/value pairs
                        reduced_dict: dict = {key: x[key] for key in self._USED_KEYS}
                        station_list.append(reduced_dict)
        # Create an indexed dictionary
        self._station_collection = dict((station['DS100'], station) for station in station_list)
        print(self._station_collection)

    def get_station_name(self, acronym: str) -> str:
        """Returns the station name for a given station acronym

        Parameters
        ----------
        acronym : str
            The acronym (DS100-Code) of the requested station

        Returns
        -------
        str
            Name of the station

        Raises
        ------
        StationNotFoundException
            If no station for the given acronym (DS100 code) could be found
        """
        if acronym in self._station_collection:
            return self._station_collection[acronym]['NAME']
        else:
            raise StationNotFoundException

    def get_station_coords(self, acronym: str) -> tuple[float, float]:
        """Return the coordinates for a given station acronym

        Parameters
        ----------
        acronym : str
            The acronym (DS100-Code) of the requested station

        Returns
        -------
        tuple[float, float]
            Coordinate tuple (longitude, latitude)

        Raises
        ------
        StationNotFoundException
            If no station for the given acronym (DS100 code) could be found
        """
        if acronym in self._station_collection:
            return self._station_collection[acronym]['Laenge'],\
                self._station_collection[acronym]['Breite']
        else:
            raise StationNotFoundException


class StationNotFoundException(Exception):
    # Custom Exception for missing requested stations
    """Requested station couldn't be found"""
    pass

File: kotlin/station-distance/src/main/kotlin/stationdistance/Stations.kt
package stationdistance

import org.apache.commons.csv.CSVFormat
import org.apache.commons.csv.CSVParser
import java.nio.file.Paths
import java.nio.file.Files


// Path to input csv file
private const val CSVPATH: String = "data/D_Bahnhof_2020_alle.CSV"

/**
 * The Station class manages the collection of stations and provides some methods for it.
 */
class Stations {
    private val stationMap: MutableMap<String, Station> = mutableMapOf()

    init {
        val reader = Files.newBufferedReader(Paths.get(CSVPATH))
        // Parse csv and drop first line (column titles)
        val csvParser = CSVParser(reader, CSVFormat.newFormat(';')).drop(1)
        for (line in csvParser) {
            // Filter on only Fernverkehrshalte (FV)
            if (line[4] == "FV") {
                // Some stations have multiple DS100 codes
                // Therefore they need to be added individually.
                line[1].split(',').forEach {
                    // Create Station instance and add it to stationMap
                    stationMap[it] = Station(ds100 = it,
                        name = line[3],
                        coordinates = Pair(line[5].replace(',', '.').toDouble(),
                            line[6].replace(',', '.').toDouble()))
                }
            }
        }
    }

    /**
     * Returns the station name for a given station DS100 code
     * @param ds100 The DS100-Code of the requested station
     * @return Name of the station
     * @throws StationNotFoundException
     */
    fun getStationName(ds100: String) : String {
        val station : Station? = stationMap[ds100]
        if (station != null) {
            return station.name
        } else {
            throw StationNotFoundException("No station with ds100 code $ds100")
        }
    }

    /**
     * Returns the coordinates for a given station DS100 code
     * @param ds100 The DS100-Code of the requested station
     * @return Coordinate pair (longitude, latitude)
     * @throws StationNotFoundException
     */
    fun getStationCoords(ds100: String) : Pair<Double, Double> {
        val station : Station? = stationMap[ds100]
        if (station != null) {
            return station.coordinates
        } else {
            throw StationNotFoundException("No station with ds100 code $ds100")
        }
    }
}

data class Station (
    val ds100: String,
    val name: String,
    val coordinates: Pair<Double, Double>
)
/** Custom Exception for missing requested stations
* Custom Exception according: https://stackoverflow.com/a/68775013
*/
class StationNotFoundException(message: String? = null,
                               cause: Throwable? = null) : Exception(message, cause) {
    constructor(cause: Throwable) : this(null, cause)
}